### PR TITLE
SYNPY-1103 multipart copy support part 2

### PR DESCRIPTION
**Pull request by jkiang13**
Part 2 of [SYNPY-1103](https://sagebionetworks.jira.com/browse/SYNPY-1103)

Introduces synapseutils.migrate to migrate the storage location of supported entities (FIles, Tables, and Folder/Project containers containing them) by copying the file handles to the new storage location.

If the migrated entity is a Folder/Project it will first be recursively indexed in a SQLite db, and then the file handles associated with Files and Tables will be migrated ordered by their Synapse id. The ordering is intended to reduce the impact of large migrations on Synapse stack deployments and is the approach described in [SYNPY-1106](https://sagebionetworks.jira.com/browse/SYNPY-1106). A sqlite db file also serves to support resuming a migration by passing in the path of an existing sqlite file generated from an aborted previous migration, and it serves to record the file handle change made during the migration.

### CODE REVIEW

File: synapseutils/migrate_functions.py [hunk 1]
```diff
@@ -0,0 +1,736 @@
...
+installation of python. Using a Python installed from a binary package or compiled from source with sqlite
+development headers available should ensure that the sqlite3 module is available.""")
+        raise
+
+
+class _MigrationStatus(Enum):
+    # an internal enum for use within the sqlite db
+    # to track the state of entities as they are indexed
+    # and then migrated
+    UNINDEXED = 0
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:47]: this isn't used

File: synapseutils/migrate_functions.py [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+                primary key(id, version),
+                foreign key (id) references entities(id)
+            );
+        """
+    )
+
+    cursor.execute(
+        """
+            create table if not exists table_entity_files (
+                id text,
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:195]: record source storage location id

File: synapseutils/migrate_functions.py [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+
+
+def migrate(
+        syn,
+        entity,
+        storage_location_id,
+        version='new',
+        db_path=None,
+        continue_on_error=False,
+):
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:248]: If run from an interactive shell ask the user before proceeding with the migration.

File: synapseutils/migrate_functions.py [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+
+    :param syn:                                 A Synapse client instance
+    :param entity:                              The entity to migrate, typically a Folder or Project
+    :param storage_location_id:                 The storage location where the file handle(s) will be migrated to
+    :param version:                             One of the following:
+                                                    'new': create a new version for entities that are migrated
+                                                    'all': migrate all entities in place by updating their file handles
+                                                    'latest: migrate the latest entity version in place by updating
+                                                            its file handle
+    :param db_path:                             a path where a SQLite database can be saved to coordinate the progress
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:260]: Force passing in the db path

File: synapseutils/migrate_functions.py [hunk 5]
```diff
@@ -0,0 +1,736 @@
...
+        entity_type = _entity_type_of(entity)
+
+        if entity_type == 'file' or entity_type == 'table':
+            cursor.execute(
+                'insert into entities (id, parent_id, type, status) values (?, ?, ?, ?)',
+                (entity_id, parent_id, entity_type, _MigrationStatus.INDEXED.value)
+            )
+
+        elif entity_type == 'folder' or entity_type == 'project':
+            children = syn.getChildren(entity_id, includeTypes=['file', 'table', 'folder'])
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:455]: Don't walk tables for now... just FileEntities

File: synapseutils/migrate_functions.py [hunk 6]
```diff
@@ -0,0 +1,736 @@
...
+    entity_id = utils.id_of(entity)
+    entity = syn.get(entity, downloadFile=False, version=version)
+    version = entity.versionNumber
+
+    existing_file_handle_id = entity.dataFileHandleId
+    existing_file_name = entity._file_handle['fileName']
+    existing_storage_location_id = entity._file_handle['storageLocationId']
+
+    if str(existing_storage_location_id) == str(storage_location_id):
+        logging.info(
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:604]: warn?

File: synapseutils/migrate_functions.py [hunk 7]
```diff
@@ -0,0 +1,736 @@
...
+    """
+
+    entity = syn.get(entity, downloadFile=False)
+    if not isinstance(entity, synapseclient.Schema):
+        raise ValueError('passed value is not a table Schema')
+
+    entity_id = utils.id_of(entity)
+
+    file_columns = []
+    for col in syn.restGET("/entity/{id}/column".format(id=entity_id))['results']:
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:674]: fstring

File: synapseclient/__main__.py [hunk 8]
```diff
@@ -843,6 +870,23 @@ def build_parser():
...
+        help='Migrate Synapse entities to a different storage location'
+    )
+    parser_migrate.add_argument('id', type=str, help='Synapse id')
+    parser_migrate.add_argument('storage_location_id', type=str, help='Synapse storage location id')
+    parser_migrate.add_argument('--version', type=str, default='new',
+                                help="""one of 'new', 'latest', 'all', or a specific version number:
+                                     new creates a new version of each entity,
+                                     latest migrates the most recent version,
+                                     all migrates all versions""")
+    parser_migrate.add_argument('--db_path', type=str,
```
Comment (jkiang13) [inline synapseclient/__main__.py:884]: no longer optional

File: synapseutils/migrate_functions.py (hunk 1 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py:47]: 0 value could be a dry run.

File: synapseutils/migrate_functions.py [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+        """
+    )
+
+    cursor.execute(
+        """
+            create table if not exists file_entity_versions (
+                id text,
+                version integer,
+                from_file_handle_id text not null,
+                to_file_handle_id text not null,
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:185]: if dry run these would be nullable

File: synapseutils/migrate_functions.py [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+        version_number = version['versionNumber']
+        try:
+            _, from_file_handle_id, to_file_handle_id = _migrate_file_version(
+                syn,
+                entity,
+                storage_location_id,
+                version_number,
+            )
+
+            mapping[version_number] = (from_file_handle_id, to_file_handle_id)
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:584]: catch the individual exceptions for versions

File: synapseutils/migrate_functions.py [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+    :returns: a 5-tuple consisting detailing the results of the migration;
+        * map of file entity id to (version number, old file handle id, new file handle id)
+        * a list of file entities that were not able to be migrated (exceptions recorded a the sqlite db)
+        * map of tables with attached file handles that were migrated,
+            entity id to (row id, old file handle id, new file handle id)
+        * a list of tables that were not able to be migrated (exceptions recorded in the sqlite db)
+        * path to the sqlite db
+    """
+
+    if db_path is None:
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:275]: validate storage location

File: synapseutils/migrate_functions.py (hunk 1 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: Removed.

File: synapseutils/migrate_functions.py (hunk 3 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py:411]: Migration now requires either interactive confirmation or using a force parameter.

File: synapseutils/migrate_functions.py (hunk 2 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: source storage location id now recorded in the index

File: synapseutils/migrate_functions.py (hunk 4 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: Required arg now.

File: synapseutils/migrate_functions.py (hunk 5 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: By default we only walk files now, but tables can be included with the optional include_table_files param.

File: synapseutils/migrate_functions.py (hunk 6 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: Removed this logging statement, already migrated items are recorded in the index.

File: synapseutils/migrate_functions.py (hunk 10 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: Flattened the sqlite db, every version can now record an exception.

File: synapseutils/migrate_functions.py (hunk 11 above)
Comment (jkiang13) [reply synapseutils/migrate_functions.py]: Destination storage location ownership now verified.

**Comment by jkiang13** [comment]
Maybe make recursion optional.

File: synapseutils/migrate_functions.py [hunk 12]
```diff
@@ -290,79 +399,183 @@ def _wait_futures(conn, cursor, futures, return_when, continue_on_error):
...
+
+def migrate_indexed_files(
+    syn: synapseclient.Synapse,
+    db_path: str,
+    create_table_snapshots=True,
+    continue_on_error=False,
+    force=False
+):
+    """
+    Migrate files previously indexed in a sqlite database at the given db_path using the separate
```
Comment (jkiang13) [inline synapseutils/migrate_functions.py:541]: Add to doc warning that if the entity changes between the time of the index and the time of migration then changes may not be reflected in the migration.

**Comment by jkiang13** [comment]
split the version in the migration table into a source and target version

**Comment by jkiang13** [comment]
warn that sqlite is single threaded should only be operating from one process at a time

File: tests/unit/synapseutils/unit_test_synapseutils_migrate.py [hunk 13]
```diff
@@ -0,0 +1,1581 @@
...
+            include_table_files,
+            continue_on_error
+        )
+
+        mock_check_indexed.assert_called_once_with(cursor, entity_id)
+        assert mock_index_file_entity.called is False
+
+    @mock.patch.object(synapseutils.migrate_functions, '_index_file_entity')
+    @mock.patch.object(synapseutils.migrate_functions, '_check_indexed')
+    def test_index_entity__file(self, mock_check_indexed, mock_index_file_entity):
```
Comment (jkiang13) [inline tests/unit/synapseutils/unit_test_synapseutils_migrate.py:636]: Add a test on an error on a specific version

**Review by marcomarasca — APPROVED** [review]
Let us merge this for now and get the changes in the next PR.

**Review by zimingd — APPROVED** [review]

**Pull request merged**